blog/content/post/2020-07-21-secure-app-coding-practices.md

```diff
@@ -0,0 +1,72 @@
+  - Security
+---
+
+When building an application, we must assume that it will be exposed to hackers at all times, but also to misuse by ordinary recipients. The danger of the first group seems obvious, but what kind of risk are other users? You must know that attempts to "cheat the system" are quite common, especially in the e-commerce industry. There is a group of users who count, for example, on errors in calculating discounts or the number of elements. You can expect that if you have an online store, there will be at least a few people who will try to use your site in such a way as to find a vulnerability. Therefore, we must be really careful when building an application. We cannot count on everyone to act according to our plan. A secure application is one in which the author assumes the opposite scenario - such that everyone wants to find a vulnerability in our system. Only such an approach, and then appropriate protection, can give us at least a substitute for security.
+
+We will try to approach the topic from a slightly easier side - what are the best practices for writing secure web applications?
+There are many of them, but here we'll cover some very basic rules to keep in mind.
+
+### Hide sensitive data
+
+This may seem obvious to you but the passwords for our accounts are so important that we should protect them as tightly as possible.
+If you make a test app and keep it in your repository you can say it doesn't matter. However, have you really not used this password somewhere? Hackers know very well that most internet users often use the exact same combination of characters. So you can expose yourself to the fact that someone will try to use the login data found, e.g. on Twitter or Google account. An even worse scenario is possible, because what if you don't change your password before the app is released to the world? Even if you make your repository private from then on, there's no guarantee that someone hasn't already saved your data. Such neglect could cost you a lot.
+First of all it is good practice to use separate channels of communication for sharing sensitive information such as a PIN or password. For example, you can use an HTTPS network connection to share encrypted data between the client and server, and then use APNS/GCM or SMS to share the PIN or token/key with the client app. That way, even if one channel of communication is compromised, the security of the overall system remains intact. Additionaly most local development and hosting environments now provide a mechanism for registering custom [environment variables](https://docs.microsoft.com/en-us/aspnet/core/security/app-secrets?view=aspnetcore-3.1&tabs=windows/).
+
+Don't forget though that keeping your server code on the repository, even without passwords, can be risky. If your script has any vulnerabilities, the hacker will be able to easily detect them, and then attack the application published, for example to download data from the database using it.
```

Comment:
This is kind of against open-source approach. When others can look into the code, they can point out those security issues, maybe even fix it. I think hiding the code makes it harder to fix than to break.